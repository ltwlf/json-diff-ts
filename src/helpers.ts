export function splitJSONPath(path: string): string[] {
    const parts: string[] = [];
    let currentPart = '';
    let inSingleQuotes = false;
    let inBrackets = 0;

    for (let i = 0; i < path.length; i++) {
        const char = path[i];

        if (char === "'" && path[i - 1] !== '\\') {
            // Toggle single quote flag if not escaped
            inSingleQuotes = !inSingleQuotes;
        } else if (char === '[' && !inSingleQuotes) {
            // Increase bracket nesting level
            inBrackets++;
        } else if (char === ']' && !inSingleQuotes) {
            // Decrease bracket nesting level
            inBrackets--;
        }

        if (char === '.' && !inSingleQuotes && inBrackets === 0) {
            // Split at period if not in quotes or brackets
            parts.push(currentPart);
            currentPart = '';
        } else {
            // Otherwise, keep adding to the current part
            currentPart += char;
        }
    }

    // Add the last part if there's any
    if (currentPart !== '') {
        parts.push(currentPart);
    }

    return parts;
}
