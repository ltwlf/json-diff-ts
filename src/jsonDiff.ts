import { difference, find, intersection, keyBy } from 'lodash-es';

interface Dictionary<T> {
  [index: string]: T;
}

type FunctionKey = (obj: any, getKeyName?: boolean) => any;

export const getTypeOfObj = (obj: any) => {
  if (typeof obj === 'undefined') {
    return 'undefined';
  }

  if (obj === null) {
    return null;
  }

  return Object.prototype.toString.call(obj).match(/^\[object\s(.*)\]$/)[1];
};

const getKey = (path: string) => {
  const left = path[path.length - 1];
  return left != null ? left : '$root';
};

const compare = (oldObj: any, newObj: any, path: any, embeddedObjKeys: any, keyPath: any) => {
  let changes: any[] = [];

  const typeOfOldObj = getTypeOfObj(oldObj);

  switch (typeOfOldObj) {
    case 'Date':
      changes = changes.concat(
        comparePrimitives(oldObj.getTime(), newObj.getTime(), path).map((x) => ({
          ...x,
          value: new Date(x.value),
          oldValue: new Date(x.oldValue)
        }))
      );
      break;
    case 'Object':
      const diffs = compareObject(oldObj, newObj, path, embeddedObjKeys, keyPath);
      if (diffs.length) {
        if (path.length) {
          changes.push({
            type: Operation.UPDATE,
            key: getKey(path),
            changes: diffs
          });
        } else {
          changes = changes.concat(diffs);
        }
      }
      break;
    case 'Array':
      changes = changes.concat(compareArray(oldObj, newObj, path, embeddedObjKeys, keyPath));
      break;
    case 'Function':
      break;
    // do nothing
    default:
      changes = changes.concat(comparePrimitives(oldObj, newObj, path));
  }

  return changes;
};

const compareObject = (oldObj: any, newObj: any, path: any, embeddedObjKeys: any, keyPath: any, skipPath = false) => {
  let k;
  let newKeyPath;
  let newPath;

  if (skipPath == null) {
    skipPath = false;
  }
  let changes: any[] = [];

  const oldObjKeys = Object.keys(oldObj);
  const newObjKeys = Object.keys(newObj);

  const intersectionKeys = intersection(oldObjKeys, newObjKeys);
  for (k of intersectionKeys) {
    newPath = path.concat([k]);
    newKeyPath = skipPath ? keyPath : keyPath.concat([k]);
    const diffs = compare(oldObj[k], newObj[k], newPath, embeddedObjKeys, newKeyPath);
    if (diffs.length) {
      changes = changes.concat(diffs);
    }
  }

  const addedKeys = difference(newObjKeys, oldObjKeys);
  for (k of addedKeys) {
    newPath = path.concat([k]);
    newKeyPath = skipPath ? keyPath : keyPath.concat([k]);
    changes.push({
      type: Operation.ADD,
      key: getKey(newPath),
      value: newObj[k]
    });
  }

  const deletedKeys = difference(oldObjKeys, newObjKeys);
  for (k of deletedKeys) {
    newPath = path.concat([k]);
    newKeyPath = skipPath ? keyPath : keyPath.concat([k]);
    changes.push({
      type: Operation.REMOVE,
      key: getKey(newPath),
      value: oldObj[k]
    });
  }
  return changes;
};

const compareArray = (oldObj: any, newObj: any, path: any, embeddedObjKeys: any, keyPath: any) => {
  const left = getObjectKey(embeddedObjKeys, keyPath);
  const uniqKey = left != null ? left : '$index';
  const indexedOldObj = convertArrayToObj(oldObj, uniqKey);
  const indexedNewObj = convertArrayToObj(newObj, uniqKey);
  const diffs = compareObject(indexedOldObj, indexedNewObj, path, embeddedObjKeys, keyPath, true);
  if (diffs.length) {
    return [
      {
        type: Operation.UPDATE,
        key: getKey(path),
        embeddedKey: typeof uniqKey === 'function' && uniqKey.length === 2 ? uniqKey(newObj[0], true) : uniqKey,
        changes: diffs
      }
    ];
  } else {
    return [];
  }
};

const getObjectKey = (embeddedObjKeys: any, keyPath: any) => {
  if (embeddedObjKeys != null) {
    const path = keyPath.join('.');
    const key = embeddedObjKeys[path];
    if (key != null) {
      return key;
    }
    for (const regex in embeddedObjKeys) {
      if (path.match(new RegExp(regex))) {
        return embeddedObjKeys[regex];
      }
    }
  }
  return undefined;
};

const convertArrayToObj = (arr: any[], uniqKey: any) => {
  let obj: any = {};
  if (uniqKey !== '$index') {
    obj = keyBy(arr, uniqKey);
  } else {
    for (let i = 0; i < arr.length; i++) {
      const value = arr[i];
      obj[i] = value;
    }
  }
  return obj;
};

const comparePrimitives = (oldObj: any, newObj: any, path: any) => {
  const changes = [];
  if (oldObj !== newObj) {
    changes.push({
      type: Operation.UPDATE,
      key: getKey(path),
      value: newObj,
      oldValue: oldObj
    });
  }
  return changes;
};

// const isEmbeddedKey = key => /\$.*=/gi.test(key)

const removeKey = (obj: any, key: any, embeddedKey: any) => {
  if (Array.isArray(obj)) {
    if (embeddedKey === '$index') {
      obj.splice(key);
      return;
    }
    const index = indexOfItemInArray(obj, embeddedKey, key);
    if (index === -1) {
      // tslint:disable-next-line:no-console
      console.warn(`Element with the key '${embeddedKey}' and value '${key}' could not be found in the array'`);
      return;
    }
    return obj.splice(index != null ? index : key, 1);
  } else {
    obj[key] = undefined;
    delete obj[key];
    return;
  }
};

const indexOfItemInArray = (arr: any[], key: any, value: any) => {
  for (let i = 0; i < arr.length; i++) {
    const item = arr[i];
    if (item && item[key] ? item[key].toString() === value.toString() : undefined) {
      return i;
    }
  }
  return -1;
};

const modifyKeyValue = (obj: any, key: any, value: any) => (obj[key] = value);

const addKeyValue = (obj: any, key: any, value: any) => {
  if (Array.isArray(obj)) {
    return obj.push(value);
  } else {
    return obj ? (obj[key] = value) : null;
  }
};

const applyLeafChange = (obj: any, change: any, embeddedKey: any) => {
  const { type, key, value } = change;
  switch (type) {
    case Operation.ADD:
      return addKeyValue(obj, key, value);
    case Operation.UPDATE:
      return modifyKeyValue(obj, key, value);
    case Operation.REMOVE:
      return removeKey(obj, key, embeddedKey);
  }
};

const applyArrayChange = (arr: any, change: any) =>
  (() => {
    const result = [];
    for (const subchange of change.changes) {
      if (subchange.value != null || subchange.type === Operation.REMOVE) {
        result.push(applyLeafChange(arr, subchange, change.embeddedKey));
      } else {
        let element;
        if (change.embeddedKey === '$index') {
          element = arr[subchange.key];
        } else {
          element = find(arr, (el) => el[change.embeddedKey].toString() === subchange.key.toString());
        }
        result.push(applyChangeset(element, subchange.changes));
      }
    }
    return result;
  })();

const applyBranchChange = (obj: any, change: any) => {
  if (Array.isArray(obj)) {
    return applyArrayChange(obj, change);
  } else {
    return applyChangeset(obj, change.changes);
  }
};

const revertLeafChange = (obj: any, change: any, embeddedKey = '$index') => {
  const { type, key, value, oldValue } = change;
  switch (type) {
    case Operation.ADD:
      return removeKey(obj, key, embeddedKey);
    case Operation.UPDATE:
      return modifyKeyValue(obj, key, oldValue);
    case Operation.REMOVE:
      return addKeyValue(obj, key, value);
  }
};

const revertArrayChange = (arr: any, change: any) =>
  (() => {
    const result = [];
    for (const subchange of change.changes) {
      if (subchange.value != null || subchange.type === Operation.REMOVE) {
        result.push(revertLeafChange(arr, subchange, change.embeddedKey));
      } else {
        let element;
        if (change.embeddedKey === '$index') {
          element = arr[+subchange.key];
        } else {
          element = find(arr, (el) => el[change.embeddedKey].toString() === subchange.key);
        }
        result.push(revertChangeset(element, subchange.changes));
      }
    }
    return result;
  })();

const revertBranchChange = (obj: any, change: any) => {
  if (Array.isArray(obj)) {
    return revertArrayChange(obj, change);
  } else {
    return revertChangeset(obj, change.changes);
  }
};

export const diff = (oldObj: any, newObj: any, embeddedObjKeys?: Dictionary<string | FunctionKey>): IChange[] =>
  compare(oldObj, newObj, [], embeddedObjKeys, []);

export const applyChangeset = (obj: any, changeset: Changeset) => {
  if (changeset) {
    changeset.forEach((change) =>
      (change.value !== null && change.value !== undefined) || change.type === Operation.REMOVE
        ? applyLeafChange(obj, change, change.embeddedKey)
        : applyBranchChange(obj[change.key], change)
    );
  }

  return obj;
};

export const revertChangeset = (obj: any, changeset: Changeset) => {
  if (changeset) {
    changeset
      .reverse()
      .forEach((change: IChange): any =>
        !change.changes ? revertLeafChange(obj, change) : revertBranchChange(obj[change.key], change)
      );
  }
  return obj;
};

export enum Operation {
  REMOVE = 'REMOVE',
  ADD = 'ADD',
  UPDATE = 'UPDATE'
}

export interface IChange {
  type: Operation;
  key: string;
  embeddedKey?: string | FunctionKey;
  value?: any | any[];
  oldValue?: any;
  changes?: IChange[];
}
export type Changeset = IChange[];

export interface IFlatChange {
  type: Operation;
  key: string;
  path: string;
  valueType: string | null;
  value?: any;
  oldValue?: any;
}

export const flattenChangeset = (
  obj: Changeset | IChange,
  path = '$',
  embeddedKey?: string | FunctionKey
): IFlatChange[] => {
  if (Array.isArray(obj)) {
    return obj.reduce((memo, change) => [...memo, ...flattenChangeset(change, path, embeddedKey)], [] as IFlatChange[]);
  } else {
    if (obj.changes || embeddedKey) {
      path = embeddedKey
        ? embeddedKey === '$index'
          ? `${path}[${obj.key}]`
          : obj.type === Operation.ADD
          ? path
          : `${path}[?(@.${embeddedKey}='${obj.key}')]`
        : (path = `${path}.${obj.key}`);
      return flattenChangeset(obj.changes || obj, path, obj.embeddedKey);
    } else {
      const valueType = getTypeOfObj(obj.value);
      return [
        {
          ...obj,
          path: valueType === 'Object' || path.endsWith(`[${obj.key}]`) ? path : `${path}.${obj.key}`,
          valueType
        }
      ];
    }
  }
};

export const unflattenChanges = (changes: IFlatChange | IFlatChange[]) => {
  if (!Array.isArray(changes)) {
    changes = [changes];
  }

  const changesArr: IChange[] = [];

  changes.forEach((change) => {
    const obj = {} as IChange;
    let ptr = obj;

    const segments = change.path.split(/([^@])\./).reduce((acc, curr, i) => {
      const x = Math.floor(i / 2);
      if (!acc[x]) {
        acc[x] = '';
      }
      acc[x] += curr;
      return acc;
    }, []);
    // $.childern[@.name='chris'].age
    // =>
    // $
    // childern[@.name='chris']
    // age

    if (segments.length === 1) {
      ptr.key = change.key;
      ptr.type = change.type;
      ptr.value = change.value;
      ptr.oldValue = change.oldValue;
      changesArr.push(ptr);
    } else {
      for (let i = 1; i < segments.length; i++) {
        const segment = segments[i];
        // check for array
        const result = /^(.+)\[\?\(@\.(.+)='(.+)'\)]$|^(.+)\[(\d+)\]/.exec(segment);
        // array
        if (result) {
          let key: string;
          let embeddedKey: string;
          let arrKey: string | number;
          if (result[1]) {
            key = result[1];
            embeddedKey = result[2];
            arrKey = result[3];
          } else {
            key = result[4];
            embeddedKey = '$index';
            arrKey = Number(result[5]);
          }
          // leaf
          if (i === segments.length - 1) {
            ptr.key = key!;
            ptr.embeddedKey = embeddedKey!;
            ptr.type = Operation.UPDATE;
            ptr.changes = [
              {
                type: change.type,
                key: arrKey!,
                value: change.value,
                oldValue: change.oldValue
              } as IChange
            ];
          } else {
            // object
            ptr.key = key;
            ptr.embeddedKey = embeddedKey;
            ptr.type = Operation.UPDATE;
            const newPtr = {} as IChange;
            ptr.changes = [
              {
                type: Operation.UPDATE,
                key: arrKey,
                changes: [newPtr]
              } as IChange
            ];
            ptr = newPtr;
          }
        } else {
          // leaf
          if (i === segments.length - 1) {
            // check if value is a primitive or object
            if (change.value !== null && change.valueType === 'Object') {
              ptr.key = segment;
              ptr.type = Operation.UPDATE;
              ptr.changes = [
                {
                  key: change.key,
                  type: change.type,
                  value: change.value
                } as IChange
              ];
            } else {
              ptr.key = change.key;
              ptr.type = change.type;
              ptr.value = change.value;
              ptr.oldValue = change.oldValue;
            }
          } else {
            // branch
            ptr.key = segment;
            ptr.type = Operation.UPDATE;
            const newPtr = {} as IChange;
            ptr.changes = [newPtr];
            ptr = newPtr;
          }
        }
      }
      changesArr.push(obj);
    }
  });
  return changesArr;
};
