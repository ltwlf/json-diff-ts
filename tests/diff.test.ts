import { diff, EmbeddedObjKeysMapType } from '../src/jsonDiff';
import * as fixtures from './__fixtures__/jsonDiff.fixture';

let oldObj: any;
let newObj: any;

beforeEach(() => {
  oldObj = fixtures.oldObj();
  newObj = fixtures.newObj();
});

describe('diff', () => {
  it('returns correct diff for objects with embedded array without specified key', () => {
    const diffs = diff(oldObj, newObj);
    expect(diffs).toMatchSnapshot();
  });

  it('returns correct diff for objects with embedded array with specified keys', () => {
    const diffs = diff(oldObj, newObj, {
      embeddedObjKeys: {
        children: 'name',
        // path can either starts with "" or "."
        '.children.subset': 'id'
      }
    });
    expect(diffs).toMatchSnapshot();
  });

  it('returns correct diff for objects with embedded array with regex keys', () => {
    const embeddedObjKeys: EmbeddedObjKeysMapType = new Map();
    embeddedObjKeys.set(/^children$/, 'name');
    embeddedObjKeys.set(/\.subset$/, 'id');

    const diffs = diff(oldObj, newObj, { embeddedObjKeys });
    expect(diffs).toMatchSnapshot();
  });

  it('returns correct diff for objects with embedded array with function keys', () => {
    const diffs = diff(oldObj, newObj, {
      embeddedObjKeys: {
        children: (obj: { name: string }) => obj.name,
        'children.subset': (obj: { id: number }) => obj.id
      }
    });
    expect(diffs).toMatchSnapshot();
  });

  it('returns correct diff for object without keys to skip', () => {
    const keyToSkip = '@_index';
    oldObj[keyToSkip] = 'This should be ignored';
    newObj['children'][1][keyToSkip] = { text: 'This whole object should be ignored' };
    const diffs = diff(oldObj, newObj, { keysToSkip: [keyToSkip] });
    // Update the snapshot with npm test -- -u if needed
    expect(diffs).toMatchSnapshot();
  });

  it('supports nested keys to skip', () => {
    const original = {
      property: {
        name: 'Paucek, Gerlach and Bernier',
        address: {
          formattedAddress: '80568 Abernathy Pine Apt. 387',
          utcOffset: 0,
          vicinity: '866 Woodside Road Apt. 534',
        }
      }
    };
    const updated = {
      property: {
        name: 'New Address',
        address: {
          formattedAddress: 'New 80568 Abernathy Pine Apt. 387',
          utcOffset: 0,
          vicinity: 'New 866 Woodside Road Apt. 534',
        }
      }
    };
    
    const diffs = diff(original, updated, { keysToSkip: ['property.address'] });
    expect(diffs).toEqual([
      {
        type: 'UPDATE',
        key: 'property',
        changes: [
          {
            type: 'UPDATE',
            key: 'name',
            value: 'New Address',
            oldValue: 'Paucek, Gerlach and Bernier'
          }
        ]
      }
    ]);
  });

  it.each(fixtures.assortedDiffs)(
    'correctly diffs $oldVal with $newVal',
    ({ oldVal, newVal, expectedReplacement, expectedUpdate }) => {
      expect(diff(oldVal, newVal, { treatTypeChangeAsReplace: true })).toEqual(expectedReplacement);
      expect(diff(oldVal, newVal, { treatTypeChangeAsReplace: false })).toEqual(expectedUpdate);
    }
  );

  it('should not include empty REMOVE operation when diffing from undefined to a value', () => {
    const value = { DBA: "New Val" };
    const valueDiff = diff(undefined, value);
    
    // Check that there's no REMOVE operation
    const removeOperation = valueDiff.find(change => change.type === 'REMOVE');
    
    expect(removeOperation).toBeUndefined();
    
    // Check that there's only an ADD operation
    expect(valueDiff.length).toBe(1);
    expect(valueDiff[0].type).toBe('ADD');
    expect(valueDiff[0].key).toBe('$root');
    expect(valueDiff[0].value).toEqual(value);
  });
  
  it('should include a REMOVE operation with value when diffing from a value to undefined', () => {
    const value = { DBA: "New Val" };
    const valueDiff = diff(value, undefined);
    
    // Check if there's a REMOVE operation with the original value
    expect(valueDiff.length).toBe(1);
    expect(valueDiff[0].type).toBe('REMOVE');
    expect(valueDiff[0].key).toBe('$root');
    expect(valueDiff[0].value).toEqual(value);
  });

  it('handles Date to string updates when treatTypeChangeAsReplace is false (issue #254)', () => {
    const d = '2025-05-28T06:40:53.284Z';
    const before = { d: new Date(d) };
    const after = { d };

    const valueDiff = diff(before, after, { treatTypeChangeAsReplace: false });

    expect(valueDiff).toEqual([
      { type: 'UPDATE', key: 'd', value: d, oldValue: new Date(d) }
    ]);
  });
});