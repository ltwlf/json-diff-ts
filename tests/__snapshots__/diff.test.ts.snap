// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`diff returns correct diff for object without keys to skip 1`] = `
[
  {
    "key": "name",
    "oldValue": "joe",
    "type": "UPDATE",
    "value": "smith",
  },
  {
    "key": "mixed",
    "type": "REMOVE",
    "value": 10,
  },
  {
    "key": "mixed",
    "type": "ADD",
    "value": "10",
  },
  {
    "changes": [
      {
        "key": "inner",
        "oldValue": 1,
        "type": "UPDATE",
        "value": 2,
      },
    ],
    "key": "nested",
    "type": "UPDATE",
  },
  {
    "key": "date",
    "oldValue": 2014-10-13T11:13:00.000Z,
    "type": "UPDATE",
    "value": 2014-10-12T11:13:00.000Z,
  },
  {
    "changes": [
      {
        "key": "2",
        "type": "ADD",
        "value": 1,
      },
    ],
    "embeddedKey": "$index",
    "key": "coins",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": "car",
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": "doll",
      },
      {
        "key": "2",
        "type": "REMOVE",
        "value": "car",
      },
    ],
    "embeddedKey": "$index",
    "key": "toys",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": undefined,
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": null,
      },
    ],
    "embeddedKey": "$index",
    "key": "pets",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "changes": [
          {
            "key": "name",
            "oldValue": "kid1",
            "type": "UPDATE",
            "value": "kid3",
          },
          {
            "key": "age",
            "oldValue": 1,
            "type": "UPDATE",
            "value": 3,
          },
          {
            "key": "subset",
            "type": "REMOVE",
            "value": [
              {
                "id": 1,
                "value": "haha",
              },
              {
                "id": 2,
                "value": "hehe",
              },
            ],
          },
        ],
        "key": "0",
        "type": "UPDATE",
      },
      {
        "changes": [
          {
            "key": "name",
            "oldValue": "kid2",
            "type": "UPDATE",
            "value": "kid1",
          },
          {
            "key": "age",
            "oldValue": 2,
            "type": "UPDATE",
            "value": 0,
          },
          {
            "key": "subset",
            "type": "ADD",
            "value": [
              {
                "id": 1,
                "value": "heihei",
              },
            ],
          },
          {
            "key": "@_index",
            "type": "ADD",
            "value": {
              "text": "This whole object should be ignored",
            },
          },
        ],
        "key": "1",
        "type": "UPDATE",
      },
      {
        "key": "2",
        "type": "ADD",
        "value": {
          "age": 2,
          "name": "kid2",
        },
      },
    ],
    "embeddedKey": "$index",
    "key": "children",
    "type": "UPDATE",
  },
  {
    "key": "age",
    "type": "REMOVE",
    "value": 55,
  },
  {
    "key": "empty",
    "type": "REMOVE",
    "value": undefined,
  },
]
`;

exports[`diff returns correct diff for objects with embedded array with function keys 1`] = `
[
  {
    "key": "name",
    "oldValue": "joe",
    "type": "UPDATE",
    "value": "smith",
  },
  {
    "key": "mixed",
    "type": "REMOVE",
    "value": 10,
  },
  {
    "key": "mixed",
    "type": "ADD",
    "value": "10",
  },
  {
    "changes": [
      {
        "key": "inner",
        "oldValue": 1,
        "type": "UPDATE",
        "value": 2,
      },
    ],
    "key": "nested",
    "type": "UPDATE",
  },
  {
    "key": "date",
    "oldValue": 2014-10-13T11:13:00.000Z,
    "type": "UPDATE",
    "value": 2014-10-12T11:13:00.000Z,
  },
  {
    "changes": [
      {
        "key": "2",
        "type": "ADD",
        "value": 1,
      },
    ],
    "embeddedKey": "$index",
    "key": "coins",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": "car",
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": "doll",
      },
      {
        "key": "2",
        "type": "REMOVE",
        "value": "car",
      },
    ],
    "embeddedKey": "$index",
    "key": "toys",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": undefined,
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": null,
      },
    ],
    "embeddedKey": "$index",
    "key": "pets",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "changes": [
          {
            "key": "age",
            "oldValue": 1,
            "type": "UPDATE",
            "value": 0,
          },
          {
            "changes": [
              {
                "changes": [
                  {
                    "key": "value",
                    "oldValue": "haha",
                    "type": "UPDATE",
                    "value": "heihei",
                  },
                ],
                "key": "1",
                "type": "UPDATE",
              },
              {
                "key": "2",
                "type": "REMOVE",
                "value": {
                  "id": 2,
                  "value": "hehe",
                },
              },
            ],
            "embeddedKey": [Function],
            "key": "subset",
            "type": "UPDATE",
          },
        ],
        "key": "kid1",
        "type": "UPDATE",
      },
      {
        "key": "kid3",
        "type": "ADD",
        "value": {
          "age": 3,
          "name": "kid3",
        },
      },
    ],
    "embeddedKey": [Function],
    "key": "children",
    "type": "UPDATE",
  },
  {
    "key": "age",
    "type": "REMOVE",
    "value": 55,
  },
  {
    "key": "empty",
    "type": "REMOVE",
    "value": undefined,
  },
]
`;

exports[`diff returns correct diff for objects with embedded array with regex keys 1`] = `
[
  {
    "key": "name",
    "oldValue": "joe",
    "type": "UPDATE",
    "value": "smith",
  },
  {
    "key": "mixed",
    "type": "REMOVE",
    "value": 10,
  },
  {
    "key": "mixed",
    "type": "ADD",
    "value": "10",
  },
  {
    "changes": [
      {
        "key": "inner",
        "oldValue": 1,
        "type": "UPDATE",
        "value": 2,
      },
    ],
    "key": "nested",
    "type": "UPDATE",
  },
  {
    "key": "date",
    "oldValue": 2014-10-13T11:13:00.000Z,
    "type": "UPDATE",
    "value": 2014-10-12T11:13:00.000Z,
  },
  {
    "changes": [
      {
        "key": "2",
        "type": "ADD",
        "value": 1,
      },
    ],
    "embeddedKey": "$index",
    "key": "coins",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": "car",
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": "doll",
      },
      {
        "key": "2",
        "type": "REMOVE",
        "value": "car",
      },
    ],
    "embeddedKey": "$index",
    "key": "toys",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": undefined,
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": null,
      },
    ],
    "embeddedKey": "$index",
    "key": "pets",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "changes": [
          {
            "key": "age",
            "oldValue": 1,
            "type": "UPDATE",
            "value": 0,
          },
          {
            "changes": [
              {
                "changes": [
                  {
                    "key": "value",
                    "oldValue": "haha",
                    "type": "UPDATE",
                    "value": "heihei",
                  },
                ],
                "key": "1",
                "type": "UPDATE",
              },
              {
                "key": "2",
                "type": "REMOVE",
                "value": {
                  "id": 2,
                  "value": "hehe",
                },
              },
            ],
            "embeddedKey": "id",
            "key": "subset",
            "type": "UPDATE",
          },
        ],
        "key": "kid1",
        "type": "UPDATE",
      },
      {
        "key": "kid3",
        "type": "ADD",
        "value": {
          "age": 3,
          "name": "kid3",
        },
      },
    ],
    "embeddedKey": "name",
    "key": "children",
    "type": "UPDATE",
  },
  {
    "key": "age",
    "type": "REMOVE",
    "value": 55,
  },
  {
    "key": "empty",
    "type": "REMOVE",
    "value": undefined,
  },
]
`;

exports[`diff returns correct diff for objects with embedded array with specified keys 1`] = `
[
  {
    "key": "name",
    "oldValue": "joe",
    "type": "UPDATE",
    "value": "smith",
  },
  {
    "key": "mixed",
    "type": "REMOVE",
    "value": 10,
  },
  {
    "key": "mixed",
    "type": "ADD",
    "value": "10",
  },
  {
    "changes": [
      {
        "key": "inner",
        "oldValue": 1,
        "type": "UPDATE",
        "value": 2,
      },
    ],
    "key": "nested",
    "type": "UPDATE",
  },
  {
    "key": "date",
    "oldValue": 2014-10-13T11:13:00.000Z,
    "type": "UPDATE",
    "value": 2014-10-12T11:13:00.000Z,
  },
  {
    "changes": [
      {
        "key": "2",
        "type": "ADD",
        "value": 1,
      },
    ],
    "embeddedKey": "$index",
    "key": "coins",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": "car",
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": "doll",
      },
      {
        "key": "2",
        "type": "REMOVE",
        "value": "car",
      },
    ],
    "embeddedKey": "$index",
    "key": "toys",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": undefined,
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": null,
      },
    ],
    "embeddedKey": "$index",
    "key": "pets",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "changes": [
          {
            "key": "age",
            "oldValue": 1,
            "type": "UPDATE",
            "value": 0,
          },
          {
            "changes": [
              {
                "changes": [
                  {
                    "key": "value",
                    "oldValue": "haha",
                    "type": "UPDATE",
                    "value": "heihei",
                  },
                ],
                "key": "1",
                "type": "UPDATE",
              },
              {
                "key": "2",
                "type": "REMOVE",
                "value": {
                  "id": 2,
                  "value": "hehe",
                },
              },
            ],
            "embeddedKey": "id",
            "key": "subset",
            "type": "UPDATE",
          },
        ],
        "key": "kid1",
        "type": "UPDATE",
      },
      {
        "key": "kid3",
        "type": "ADD",
        "value": {
          "age": 3,
          "name": "kid3",
        },
      },
    ],
    "embeddedKey": "name",
    "key": "children",
    "type": "UPDATE",
  },
  {
    "key": "age",
    "type": "REMOVE",
    "value": 55,
  },
  {
    "key": "empty",
    "type": "REMOVE",
    "value": undefined,
  },
]
`;

exports[`diff returns correct diff for objects with embedded array without specified key 1`] = `
[
  {
    "key": "name",
    "oldValue": "joe",
    "type": "UPDATE",
    "value": "smith",
  },
  {
    "key": "mixed",
    "type": "REMOVE",
    "value": 10,
  },
  {
    "key": "mixed",
    "type": "ADD",
    "value": "10",
  },
  {
    "changes": [
      {
        "key": "inner",
        "oldValue": 1,
        "type": "UPDATE",
        "value": 2,
      },
    ],
    "key": "nested",
    "type": "UPDATE",
  },
  {
    "key": "date",
    "oldValue": 2014-10-13T11:13:00.000Z,
    "type": "UPDATE",
    "value": 2014-10-12T11:13:00.000Z,
  },
  {
    "changes": [
      {
        "key": "2",
        "type": "ADD",
        "value": 1,
      },
    ],
    "embeddedKey": "$index",
    "key": "coins",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": "car",
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": "doll",
      },
      {
        "key": "2",
        "type": "REMOVE",
        "value": "car",
      },
    ],
    "embeddedKey": "$index",
    "key": "toys",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "key": "0",
        "type": "REMOVE",
        "value": undefined,
      },
      {
        "key": "1",
        "type": "REMOVE",
        "value": null,
      },
    ],
    "embeddedKey": "$index",
    "key": "pets",
    "type": "UPDATE",
  },
  {
    "changes": [
      {
        "changes": [
          {
            "key": "name",
            "oldValue": "kid1",
            "type": "UPDATE",
            "value": "kid3",
          },
          {
            "key": "age",
            "oldValue": 1,
            "type": "UPDATE",
            "value": 3,
          },
          {
            "key": "subset",
            "type": "REMOVE",
            "value": [
              {
                "id": 1,
                "value": "haha",
              },
              {
                "id": 2,
                "value": "hehe",
              },
            ],
          },
        ],
        "key": "0",
        "type": "UPDATE",
      },
      {
        "changes": [
          {
            "key": "name",
            "oldValue": "kid2",
            "type": "UPDATE",
            "value": "kid1",
          },
          {
            "key": "age",
            "oldValue": 2,
            "type": "UPDATE",
            "value": 0,
          },
          {
            "key": "subset",
            "type": "ADD",
            "value": [
              {
                "id": 1,
                "value": "heihei",
              },
            ],
          },
        ],
        "key": "1",
        "type": "UPDATE",
      },
      {
        "key": "2",
        "type": "ADD",
        "value": {
          "age": 2,
          "name": "kid2",
        },
      },
    ],
    "embeddedKey": "$index",
    "key": "children",
    "type": "UPDATE",
  },
  {
    "key": "age",
    "type": "REMOVE",
    "value": 55,
  },
  {
    "key": "empty",
    "type": "REMOVE",
    "value": undefined,
  },
]
`;
